File: backend/Models/Lab.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

public class Lab
{
    [BsonId]
    public int Id { get; set; }

    [BsonElement("Lab_Code")]
    public required string LabCode { get; set; }

    [BsonElement("Lab_Name")]
    public required string LabName { get; set; }

    [BsonElement("Description")]
    public string? Description { get; set; }

    [BsonElement("PPE")]
    public List<int> PPE { get; set; } = [];

    [BsonElement("Instructors")]
    public List<int> Instructors { get; set; } = [];

    [BsonElement("Students")]
    public List<int> Students { get; set; } = [];

    [BsonElement("Schedule")]
    public required List<Schedule> Schedule { get; set; }

    [BsonElement("Report")]
    public string? Report { get; set; }

    [BsonElement("SemesterID")]
    public int SemesterID { get; set; } = 0;

    [BsonElement("EndLab")]
    public bool EndLab { get; set; } = false;

    [BsonElement("Room")]
    public required string Room { get; set; }

    [BsonElement("Announcements")]
    public List<Announcement> Announcements { get; set; } = [];

    [BsonElement("Started")]
    public bool Started { get; set; } = false;

}

File: frontend/src/types/lab.ts
import { Announcement } from "./announcements";

export interface Lab {
	id: number;
	labCode: string;
	labName: string;
	description: string;
	ppe: number[];
	instructors: number[];
	students: number[];
	schedule: Schedule[];
	report: string;
	semesterID: number;
	endLab: boolean;
	room: string;
	announcements: Announcement[];
	started: boolean;
}

export interface LabDto {
	labCode: string;
	labName: string;
	description: string | null;
	ppe: number[];
	schedule: Schedule[];
	semesterID: number | null;
	room: string;
}

export interface CreateLabDto {
	lab: LabDto;
	instructor_Emails: string[];
	student_Emails: string[];
}

type DayOfWeek =
	| "Sunday"
	| "Monday"
	| "Tuesday"
	| "Wednesday"
	| "Thursday"
	| "Friday"
	| "Saturday";

export interface Schedule {
	dayOfWeek: DayOfWeek;
	startTime: string;
	endTime: string;
}

export interface UpdateLabDto
	extends Partial<
		Omit<
			Lab,
			| "id"
			| "ppe"
			| "instructors"
			| "students"
			| "report"
			| "announcements"
			| "started"
			| "endLab"
		>
	> {}

export interface LabAnalytics {
	total_attendance: number;
	total_attendance_bytime: number[];
	total_ppe_compliance: number;
	total_ppe_compliance_bytime: number[];
	ppe_compliance: Record<string, number>;
	ppe_compliance_bytime: Record<string, number[]>;
	people: Array<{
		name: string;
		attendance_percentage: number;
		ppE_compliance: Record<string, number>;
	}>;
	people_bytime: Array<{
		name: string;
		attendance_percentage: number[];
		ppE_compliance: Record<string, number[]>;
	}>;
}
